controller/local/controller.go

```diff
@@ -36,38 +40,63 @@ func (b *localController) Build(ctx context.Context, options controllerapi.Build
 	return b.ref, resp, nil
 }
 
-func (b *localController) Invoke(ctx context.Context, ref string, cfg controllerapi.ContainerConfig, ioIn io.ReadCloser, ioOut io.WriteCloser, ioErr io.WriteCloser) error {
+func (b *localController) ListProcesses(ctx context.Context, ref string) (infos []*controllerapi.ProcessInfo, retErr error) {
 	if ref != b.ref {
-		return errors.Errorf("unknown ref %q", ref)
+		return nil, errors.Errorf("unknown ref %q", ref)
 	}
-	if b.resultCtx == nil {
-		return errors.New("no build result is registered")
+	return b.processes.ListProcesses(), nil
+}
+
+func (b *localController) DisconnectProcess(ctx context.Context, ref, pid string) error {
+	if ref != b.ref {
+		return errors.Errorf("unknown ref %q", ref)
 	}
-	ccfg := build.ContainerConfig{
-		ResultCtx:  b.resultCtx,
-		Entrypoint: cfg.Entrypoint,
-		Cmd:        cfg.Cmd,
-		Env:        cfg.Env,
-		Tty:        cfg.Tty,
-		Stdin:      ioIn,
-		Stdout:     ioOut,
-		Stderr:     ioErr,
+	return b.processes.DeleteProcess(pid)
+}
+
+func (b *localController) cancelRunningProcesses() {
+	b.processes.CancelRunningProcesses()
+}
+
+func (b *localController) Invoke(ctx context.Context, ref string, pid string, cfg controllerapi.InvokeConfig, ioIn io.ReadCloser, ioOut io.WriteCloser, ioErr io.WriteCloser) error {
+	if ref != b.ref {
+		return errors.Errorf("unknown ref %q", ref)
 	}
-	if !cfg.NoUser {
-		ccfg.User = &cfg.User
+
+	proc, ok := b.processes.Get(pid)
+	if !ok {
+		// Start a new process.
+		if b.resultCtx == nil {
+			return errors.New("no build result is registered")
+		}
+		var err error
+		proc, err = b.processes.StartProcess(pid, b.resultCtx, &cfg)
+		if err != nil {
+			return err
+		}
 	}
-	if !cfg.NoCwd {
-		ccfg.Cwd = &cfg.Cwd
+
+	// Attach containerIn to this process
+	ioCancelledCh := make(chan struct{})
+	proc.ForwardIO(&ioset.In{Stdin: ioIn, Stdout: ioOut, Stderr: ioErr}, func() { close(ioCancelledCh) })
+
+	select {
+	case <-ioCancelledCh:
+		return errors.Errorf("io cancelled")
+	case err := <-proc.Done():
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
 	}
-	return build.Invoke(ctx, ccfg)
 }
 
 func (b *localController) Kill(context.Context) error {
 	return nil // nop
```

Comment:
We probably want a call to `b.cancelRunningProcesses()` here as well?

Reply:
Thank you for the comment. Added b.cancelRunningProcesses()